krabcake/rs_hello/src/lib.rs

```diff
@@ -815,20 +842,21 @@ pub extern "C" fn rs_trace_put(put_offset: vg_ulong, data: vg_ulong, shadow_data
                 STACKS.get_stack_dbg_id_or_assign(data as vg_addr),
                 shadow_data,
             );
-            TRACKED_GREGS.push((put_offset, Tag(shadow_data as u64)));
+            TRACKED_GREGS.push((put_offset, Tag::Id(shadow_data as u64)));
```

Comment:
All these places where you do `Tag::Id(shadow_data as u64)` are immediately suspect.

They should probably all be replaced with the helper method that decodes a Tag from a given vg_ulong.